REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [1]:
import pandas as pd

In [45]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv', dtype={"date":str})

In [3]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [4]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [5]:
bus.shape[0]

print("There are", bus.shape[0], "records.")

There are 6315 records.


#### Question 2b: How many unique business IDs are there?  

In [6]:
bus.groupby('name').agg({'business_id': 'count'})

print("There are",bus.groupby('name').agg({'business_id': 'count'}).shape[0],"unique business IDs.")

There are 5692 unique business IDs.


#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [7]:
com_bus_names = bus['name'].value_counts().head(5)

print("These are the 5 most common businesses and their count in San Francisco:",com_bus_names)

These are the 5 most common businesses and their count in San Francisco: name
STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: count, dtype: int64


## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [144]:
#when I look up a specific entry it reads as string
postal_code_type = type(bus.loc[12,'postal_code'])
postal_code_type

str

In [145]:
#however the overall column seems to store as object
bus.dtypes

business_id       int64
name             object
address          object
city             object
state            object
postal_code      object
latitude        float64
longitude       float64
phone_number     object
dtype: object

In [9]:
print("Zip code values are stored as",bus['postal_code'].dtype,"data type.")

Zip code values are stored as object data type.


#### Question 3b

What are the unique values of postal_code?

In [10]:
bus['postal_code'].unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [11]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [24]:
bus_valid = bus[bus['postal_code'].isin(validZip)]
bus_valid

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              31      NORMAN'S ICE CREAM AND FREEZES   
4              45                 CHARLIE'S DELI CAFE   
...           ...                                 ...   
6310        90234                ACCESS SFUSD THE ARC   
6311        90236                      PHO DONG HUONG   
6312        90237                    WIN'S RESTAURANT   
6313        90238                    USA POWER MARKET   
6314        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3              2801 LEAVENWORTH ST   San Francisco    CA       94133   
4                   3202 FOLSOM ST   San Francisco    CA       94110   
...                             ...            ...   ...         ...   
6310                1520 HOWARD ST   San Francisco    CA       94103   
6311             667 MONTEREY BLVD   San Francisco    CA       94127   
6312               3040 TARAVAL ST   San Francisco    CA       94116   
6313                1524 OCEAN AVE   San Francisco    CA       94112   
6314               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  
0     37.791116 -122.403816  +14154217044  
1     37.786848 -122.421547  +14157763262  
2     37.792888 -122.403135  +14156779494  
3     37.807155 -122.419004           NaN  
4     37.747114 -122.413641  +14156415051  
...         ...         ...           ...  
6310        NaN         NaN           NaN  
6311        NaN         NaN  +14154945048  
6312        NaN         NaN  +14158606036  
6313        NaN         NaN  +14153331878  
6314        NaN         NaN  +14158881393  

[5999 rows x 9 columns]

In [26]:
bus_valid['name'].value_counts()

name
STARBUCKS COFFEE                   71
PEET'S COFFEE & TEA                24
WALGREENS                          11
SAN FRANCISCO SOUP COMPANY         11
LEE'S DELI                         10
                                   ..
NORDSTROM ESPRESSO BAR              1
THE MARSH CAFE                      1
ONO HAWAIIN BBQ                     1
SAN FRANCISCO PUBLIC MONTESSORI     1
ALBA RAY'S                          1
Name: count, Length: 5438, dtype: int64

In [27]:
print("There are",len(bus_valid['name'].value_counts()),"unique businesses in this new dataframe.")

There are 5438 unique businesses in this new dataframe.


## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [28]:
bus_valid[pd.isnull(bus_valid['longitude'])]

business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Francisco    CA       94110       NaN        NaN  +14156481034  
210   San Francisco    CA       94133       NaN        NaN  +14153979919  
760   San Francisco    CA       94133       NaN        NaN  +14157715687  
826   San Francisco    CA       94102       NaN        NaN  +14153975397  
1100  San Francisco    CA       94109       NaN        NaN  +14154743482  
...             ...   ...         ...       ...        ...           ...  
6310  San Francisco    CA       94103       NaN        NaN           NaN  
6311  San Francisco    CA       94127       NaN        NaN  +14154945048  
6312  San Francisco    CA       94116       NaN        NaN  +14158606036  
6313  San Francisco    CA       94112       NaN        NaN  +14153331878  
6314  San Francisco    CA       94110       NaN        NaN  +14158881393  

[2483 rows x 9 columns]

In [29]:
miss_long = bus_valid.groupby(pd.isnull(bus_valid['longitude'])).agg({'business_id': 'count'}).loc[True, 'business_id']
miss_long

2483

In [17]:
print("There are",miss_long,"businesses missing longitude values among businesses with valid zipcodes.")

There are 2483 businesses missing longitude values among businesses with valid zipcodes.


#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [102]:
null_lon_bus_valid = bus_valid[pd.isnull(bus_valid['longitude'])]

null_lon = []
for i in validZip:
    count = null_lon_bus_valid[null_lon_bus_valid['postal_code'] == i].shape[0]
    null_lon.append(count)
    
not_null_lon_bus_valid = bus_valid[pd.notnull(bus_valid['longitude'])]

not_null_lon = []
for i in validZip:
    count = not_null_lon_bus_valid[not_null_lon_bus_valid['postal_code'] == i].shape[0]
    not_null_lon.append(count)
        
new_df = pd.DataFrame({
    "postal_code": validZip,
    "null_lon" : null_lon,
    "not_null_lon" : not_null_lon
})
new_df

postal_code  null_lon  not_null_lon
0        94102       196           262
1        94103       252           306
2        94104        59            74
3        94105       100           126
4        94107       247           204
5        94108        82           142
6        94109       146           237
7        94110       251           334
8        94111       115           167
9        94112        59           131
10       94114       108           122
11       94115        76           162
12       94116        35            58
13       94117        67           130
14       94118       100           143
15       94121        55           106
16       94122       123           147
17       94123        53           116
18       94124        88            76
19       94127        28            43
20       94131        14            34
21       94132        65            69
22       94133       135           282
23       94134        29            45

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [118]:
pd.crosstab(new_df['null_lon'], new_df['not_null_lon'], margins=True).head(5)

new_df['prop_miss'] = new_df['null_lon'] / (new_df['null_lon']+new_df['not_null_lon'])

top5_miss = new_df.sort_values(by='prop_miss', ascending=False).head(5)

print("The top five zipcodes in descending order of proportion missing longitude are")
print(top5_miss[['postal_code', 'prop_miss']])

The top five zipcodes in descending order of proportion missing longitude are
   postal_code  prop_miss
4        94107   0.547672
18       94124   0.536585
21       94132   0.485075
10       94114   0.469565
16       94122   0.455556


# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [82]:
rows = ins.shape[0]
rows

15430

In [70]:
ins.groupby('business_id').agg({'score': 'count'})

score
business_id       
10               3
19               3
24               4
31               3
45               3
...            ...
89515            1
89701            1
89790            1
89803            1
89900            1

[5730 rows x 1 columns]

In [85]:
unique_bus = len(ins['business_id'].unique())
unique_bus

5730

In [86]:
if rows == unique_bus:
    print("There is only one inspection record per business.")
else:
    print("There are multiple inspection records for some businesses.")

There are multiple inspection records for some businesses.


#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [150]:
#again, a specific entry shows string
type_type = type(ins.loc[12,'type'])

type_type

str

In [126]:
#but overall shows object using .dtype/s
ins.dtypes
print("Type is an",ins['type'].dtype,"type and takes on reference values.")

Type is an object type and takes on reference values.


In [137]:
typecounts = ins['type'].value_counts()

print("Type has",typecounts['routine'],"ocurrences for routine and",typecounts['complaint'],"for complaint.")

Type has 15429 ocurrences for routine and 1 for complaint.


In [177]:
validType = ['routine']
ins2 = ins[ins['type'].isin(validType)]
ins2

business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15429 rows x 4 columns]

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [178]:
ins2['date'] = pd.to_datetime(ins2['date'])
ins2['year'] = ins2['date'].dt.year
ins2['year']

/var/folders/19/y3bfrshn7gn6k9nck85t2kjr0000gn/T/ipykernel_52609/1173944804.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ins2['date'] = pd.to_datetime(ins2['date'])
/var/folders/19/y3bfrshn7gn6k9nck85t2kjr0000gn/T/ipykernel_52609/1173944804.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ins2['year'] = ins2['date'].dt.year


0        2016
1        2014
2        2014
3        2016
4        2014
         ... 
15425    2016
15426    2016
15427    2016
15428    2016
15429    2016
Name: year, Length: 15429, dtype: int32

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [179]:
oldest = ins2['year'].min()
recent = ins2['year'].max()

print(oldest)
print(recent)

print(oldest,"is the earliest year and", recent, "is the latest year.")

2013
2016
2013 is the earliest year and 2016 is the latest year.


In [180]:
pd.crosstab(ins2['year'], ins2['type'], margins=True)

type  routine    All
year                
2013       38     38
2014     5629   5629
2015     3923   3923
2016     5839   5839
All     15429  15429

In [187]:
validYear = [2014, 2015, 2016]
ins3 = ins2[ins2['year'].isin(validYear)]
ins3

business_id  score       date     type  year
0               10     82 2016-05-03  routine  2016
1               10     94 2014-07-29  routine  2014
2               10     92 2014-01-14  routine  2014
3               19     94 2016-05-13  routine  2016
4               19     94 2014-11-10  routine  2014
...            ...    ...        ...      ...   ...
15425        89515     98 2016-11-03  routine  2016
15426        89701    100 2016-12-06  routine  2016
15427        89790     96 2016-11-29  routine  2016
15428        89803    100 2016-12-15  routine  2016
15429        89900     93 2016-12-06  routine  2016

[15391 rows x 5 columns]

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [192]:
y2016=ins3[ins3['year'] == 2016]
y2016

business_id  score       date     type  year
0               10     82 2016-05-03  routine  2016
3               19     94 2016-05-13  routine  2016
6               24     98 2016-10-05  routine  2016
7               24     96 2016-03-11  routine  2016
10              31     84 2016-08-16  routine  2016
...            ...    ...        ...      ...   ...
15425        89515     98 2016-11-03  routine  2016
15426        89701    100 2016-12-06  routine  2016
15427        89790     96 2016-11-29  routine  2016
15428        89803    100 2016-12-15  routine  2016
15429        89900     93 2016-12-06  routine  2016

[5839 rows x 5 columns]

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [200]:
new = pd.merge(bus, y2016, how="outer", on=["business_id"])
new.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4           31      NORMAN'S ICE CREAM AND FREEZES   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
4           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   

    longitude  phone_number  score       date     type    year  
0 -122.403816  +14154217044   82.0 2016-05-03  routine  2016.0  
1 -122.421547  +14157763262   94.0 2016-05-13  routine  2016.0  
2 -122.403135  +14156779494   98.0 2016-10-05  routine  2016.0  
3 -122.403135  +14156779494   96.0 2016-03-11  routine  2016.0  
4 -122.419004           NaN   84.0 2016-08-16  routine  2016.0

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [207]:
clipped = new[['name','address','score']]
low20 = clipped.sort_values(by='score', ascending=True).head(20)
print("Below are the 20 lowest rated businesses by name, address, and score",low20)

Below are the 20 lowest rated businesses by name, address, and score                              name              address  score
4892      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    52.0
7324                    POKI TIME     2101 LOMBARD ST    55.0
6281                 CRAZY PEPPER   2257 SAN JOSE AVE    55.0
4873                   UNCLE CAFE       65 WAVERLY PL    55.0
3008             HING WANG BAKERY        339 JUDAH ST    56.0
5627                   GOLDEN WOK   295 B ORIZABA AVE    56.0
5383                   TOP'S CAFE  2460 SAN BRUNO AVE    57.0
6067    L & G VIETNAMESE SANDWICH         602 EDDY ST    57.0
836                  RED A BAKERY      634 CLEMENT ST    57.0
4209      SAN TUNG RESTAURANT LLC      1031 IRVING ST    57.0
549           MA'S DIM SUM & CAFE      1315 POWELL ST    57.0
4903  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    57.0
6645    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    58.0
624           JOANIE'S DINER INC.   1329 COLUMBUS AVE    59.0
6

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!